### PR TITLE
update c-lib to upstream

### PR DESCRIPTION
**Pull request by ordian**
I've updated the c-lib to f5c8a00, build.rs to upstream except for disabling `USE_EXTERNAL_DEFAULT_CALLBACKS` (https://github.com/paritytech/rust-secp256k1/compare/master...paritytech:ao-update-c-lib-to-upstream?expand=1#diff-a7b0a2dee0126cddf994326e705a91eaR103) (link time errors otherwise).

The benchmarks on parity-ethereum shows 33% improvement for `verify_block_unordered`
```
verify_block_unordered  time:   [17.404 ms 17.558 ms 17.720 ms]                                    
                        change: [-35.612% -35.041% -34.470%] (p = 0.00 < 0.05)
                        Performance has improved.
```

### CODE REVIEW

File: depend/secp256k1/src/ext.c [hunk 1]
```diff
@@ -24,7 +24,7 @@ int secp256k1_ecdh_raw(const secp256k1_context* ctx, unsigned char *result, cons
 		ret = 0;
 	else
 	{
-		secp256k1_ecmult_const(&res, &pt, &s);
+		secp256k1_ecmult_const(&res, &pt, &s, 256);
```
Comment (ordian) [inline depend/secp256k1/src/ext.c:27]: this needs to be verified

File: src/lib.rs [hunk 2]
```diff
@@ -329,7 +329,7 @@ impl fmt::Display for Error {
 }
 
 impl error::Error for Error {
-    fn cause(&self) -> Option<&error::Error> { None }
+    fn cause(&self) -> Option<&dyn error::Error> { None }
```
Comment (ordian) [inline src/lib.rs:332]: warning fix

File: build.rs [hunk 3]
```diff
@@ -79,30 +79,59 @@ fn setup_android(config: &mut cc::Build) {
...
-		setup_android(&mut base_config);
-	}
-
-	base_config.flag("-g")
+		.include("depend/secp256k1/src")
+		.debug(true)
+		.flag_if_supported("-Wno-unused-function") // some ecmult stuff is defined but not used upstream
+		.define("SECP256K1_BUILD", Some("1"))
+		// .define("USE_EXTERNAL_DEFAULT_CALLBACKS", Some("1"))
+		.define("ECMULT_WINDOW_SIZE", Some("15"))
```
Comment (ordian) [inline build.rs]: this could increase memory usage, needs to be checked

File: build.rs [hunk 4]
```diff
@@ -79,30 +79,59 @@ fn setup_android(config: &mut cc::Build) {
...
-	if target.contains("android") {
-		setup_android(&mut base_config);
-	}
-
-	base_config.flag("-g")
+		.include("depend/secp256k1/src")
+		.debug(true)
+		.flag_if_supported("-Wno-unused-function") // some ecmult stuff is defined but not used upstream
+		.define("SECP256K1_BUILD", Some("1"))
+		// .define("USE_EXTERNAL_DEFAULT_CALLBACKS", Some("1"))
```
Comment (dvdplm) [inline build.rs]: Intentional? If yes, can you leave a comment explaining why (and what conditions are necessary to re-enable it, if that's what we want to do some day).

File: build.rs [hunk 5]
```diff
@@ -79,30 +79,59 @@ fn setup_android(config: &mut cc::Build) {
...
-		.include("depend/secp256k1/src");
-
-	let target = env::var("TARGET").expect("TARGET env variable is set by cargo; qed");
-	if target.contains("android") {
-		setup_android(&mut base_config);
-	}
-
-	base_config.flag("-g")
+		.include("depend/secp256k1/src")
+		.debug(true)
```
Comment (dvdplm) [inline build.rs:100]: This doesn't mean the same thing as in Rust I take it? But what does it mean? Includes symbols?

File: build.rs [hunk 6]
```diff
@@ -79,30 +79,59 @@ fn setup_android(config: &mut cc::Build) {
...
+		.flag_if_supported("-Wno-unused-function") // some ecmult stuff is defined but not used upstream
+		.define("SECP256K1_BUILD", Some("1"))
+		// .define("USE_EXTERNAL_DEFAULT_CALLBACKS", Some("1"))
+		.define("ECMULT_WINDOW_SIZE", Some("15"))
+		.define("ECMULT_GEN_PREC_BITS", Some("8"))
```
Comment (dvdplm) [inline build.rs]: Where do these constants come from?

File: depend/secp256k1/src/ext.c (hunk 1 above)
Comment (dvdplm) [reply depend/secp256k1/src/ext.c:27]: Can you elaborate?

Comment (ordian) [reply depend/secp256k1/src/ext.c:27]: ext.c is not present in upstream (our extention), so we need to make sure 256 is the right size here

File: build.rs (hunk 5 above)
Comment (ordian) [reply build.rs:100]: we used "-g" flag previously, I guess this should mean the same

File: build.rs (hunk 6 above)
Comment (ordian) [reply build.rs]: this is the maximum allowed value, I'm testing different params now

File: depend/secp256k1/Makefile.am [hunk 7]
```diff
@@ -172,10 +178,6 @@ if ENABLE_MODULE_ECDH
 include src/modules/ecdh/Makefile.am.include
 endif
 
-if ENABLE_MODULE_SCHNORR
```
Comment (ordian) [inline depend/secp256k1/Makefile.am:175]: seems like schnorr support was removed, do we use it though?

File: depend/secp256k1/src/ext.c (hunk 1 above)
Comment (dvdplm) [reply depend/secp256k1/src/ext.c:27]: /cc @svyatonik PTAL?

File: build.rs (hunk 5 above)
Comment (dvdplm) [reply build.rs:100]: Hmm, so what does `-g`mean?

File: depend/secp256k1/Makefile.am (hunk 7 above)
Comment (dvdplm) [reply depend/secp256k1/Makefile.am:175]: afaik we do not, but maybe @svyatonik can help with that too?

File: build.rs (hunk 5 above)
Comment (ordian) [reply build.rs:100]: That's the same as `debug = true` in rust, adds debug symbols to the build (doesn't downgrade the perf much)

File: build.rs [hunk 8]
```diff
@@ -100,16 +100,19 @@ fn main() {
...
+		// Allowed values are: 2, 4, and 8 (tuned for best perf)
+		.define("ECMULT_GEN_PREC_BITS", Some("4"))
 		// TODO these three should be changed to use libgmp, at least until secp PR 290 is merged
 		.define("USE_NUM_NONE", Some("1"))
 		.define("USE_FIELD_INV_BUILTIN", Some("1"))
 		.define("USE_SCALAR_INV_BUILTIN", Some("1"))
 		.define("USE_ENDOMORPHISM", Some("1"))
 		.define("ENABLE_MODULE_ECDH", Some("1"))
-		.define("ENABLE_MODULE_SCHNORR", Some("1"))
+        // SCHNORR support was removed in the upstream
```
Comment (dvdplm) [inline build.rs]: Funny whitespace

File: depend/secp256k1/Makefile.am (hunk 7 above)
Comment (svyatonik) [reply depend/secp256k1/Makefile.am:175]: Ehhh...The only thing I can guarantee is that we don't use it in blockchain clients - parity-ethereum (and specifically in SecretStore), parity-bitcoin, parity-zcash (and obviously in substrate). Can't tell about the the rest...

File: depend/secp256k1/src/ext.c (hunk 1 above)
Comment (svyatonik) [reply depend/secp256k1/src/ext.c:27]: Not sure, if I am the right person to ask, but imo that's the correct change :)

**Pull request merged**